lib/model.js

```diff
@@ -3736,33 +3736,81 @@ Model.bulkWrite = function(ops, options, callback) {
     options = null;
   }
   options = options || {};
+  const ordered = options.ordered == null ? true : options.ordered;
 
   const validations = ops.map(op => castBulkWrite(this, op, options));
 
   callback = this.$handleCallbackError(callback);
   return this.db.base._promiseOrCallback(callback, cb => {
     cb = this.$wrapCallback(cb);
-    each(validations, (fn, cb) => fn(cb), error => {
-      if (error) {
-        return cb(error);
-      }
+    if (ordered) {
+      each(validations, (fn, cb) => fn(cb), error => {
+        if (error) {
+          return cb(error);
+        }
 
-      if (ops.length === 0) {
-        return cb(null, getDefaultBulkwriteResult());
-      }
+        if (ops.length === 0) {
+          return cb(null, getDefaultBulkwriteResult());
+        }
 
-      try {
-        this.$__collection.bulkWrite(ops, options, (error, res) => {
-          if (error) {
-            return cb(error);
+        try {
+          this.$__collection.bulkWrite(ops, options, (error, res) => {
+            if (error) {
+              return cb(error);
+            }
+
+            cb(null, res);
+          });
+        } catch (err) {
+          return cb(err);
+        }
+      });
+
+      return;
+    }
+
+    let remaining = validations.length;
+    let validOps = [];
+    let validationErrors = [];
+    for (let i = 0; i < validations.length; ++i) {
+      validations[i]((err) => {
+        if (err == null) {
+          validOps.push(i);
+        } else {
+          validationErrors.push({ index: i, error: err });
+        }
+        if (--remaining <= 0) {
+          completeUnorderedValidation.call(this);
+        }
+      });
+    }
+
+    validationErrors = validationErrors.
+      sort((v1, v2) => v1.index - v2.index).
+      map(v => v.error);
+
+    function completeUnorderedValidation() {
+      validOps.sort();
+      validOps = validOps.map(index => ops[index]);
```

Comment:
for consistency, why not combine it into one statement (like the above `validationErrors`)?